[caffe2][IGiOS] Fix exhaustive switches

NSProgrammer opened the pull request:
Summary:
*Exhaustive Switches*, i.e. using `-Wswitch-enum` and `-Wswitch-default`, are the ideal default for all C language code at Meta.  This ensures all enum values have an explicit case in the `switch` AND `default:` is present to handle any integer value that is not enumerated -- preventing a regular issue facing C language `switch` statements that do not have a `default:` which is either unexpected or (worse) undefined behavior.

This diff is part of the migration of parts of the codebase to adopting *Exhaustive Switches* by default.

This diff will focus on `caffe2` headers that can be used by codebases that have *Exhaustive Switches* enabled.  `switch` statements in headers will be wrapped with macros that explicitly set "mode" of each switch (*Exhaustive* or *Flexible*).  Then, each *Exhaustive* `switch` statement will be updated to have a `default` in place.

release notes: devx

Test Plan:
CI Pass

Rollback Plan:

Reviewed By: aary

Differential Revision: D77051174

zou3519 reviewed (commented):
turning on exhaustive switch seems reasonable. A lot of the tests seem like they're failing though. I'm not good enough with C++-isms to review this. Maybe @swolchok or @malfet ?

malfet commented on c10/util/Switch.h:
```diff
@@ -0,0 +1,85 @@
+*/
+
+#if defined(__has_warning) && (defined(__GNUC__) || defined(__clang__))
+#define _C10_LANG_DIAGNOSTIC_PUSH _Pragma("GCC diagnostic push")
```
Please do not introduce new macros, but rather reuse existing ones

NSProgrammer committed:
[caffe2][IGiOS] Fix exhaustive switches (#156832)

Summary:

*Exhaustive Switches*, i.e. using `-Wswitch-enum` and `-Wswitch-default`, are the ideal default for all C language code at Meta.  This ensures all enum values have an explicit case in the `switch` AND `default:` is present to handle any integer value that is not enumerated -- preventing a regular issue facing C language `switch` statements that do not have a `default:` which is either unexpected or (worse) undefined behavior.

This diff is part of the migration of parts of the codebase to adopting *Exhaustive Switches* by default.

This diff will focus on `caffe2` headers that can be used by codebases that have *Exhaustive Switches* enabled.  `switch` statements in headers will be wrapped with macros that explicitly set "mode" of each switch (*Exhaustive* or *Flexible*).  Then, each *Exhaustive* `switch` statement will be updated to have a `default` in place.

release notes: devx
topic: not user facing

Test Plan:
CI Pass

Rollback Plan:

Differential Revision: D77051174
